lib/chef/resource/osx_profile.rb

```diff
@@ -42,8 +43,241 @@ class OsxProfile < Chef::Resource
       property :identifier, String,
         description: "Use to specify the identifier for the profile, such as com.company.screensaver."
 
-      property :path, String,
-        description: "The path to write the profile to disk before loading it."
+      # this is not a property it is necessary for the tempfile this resource uses to work (FIXME: this is terrible)
+      #
+      # @api private
+      #
+      def path(path = nil)
+        @path ||= path
+        @path
+      end
+
+      action_class do
+        def load_current_resource
+          @current_resource = Chef::Resource::OsxProfile.new(new_resource.name)
+          current_resource.profile_name(new_resource.profile_name)
+
+          if new_profile_hash
+            new_profile_hash["PayloadUUID"] = config_uuid(new_profile_hash)
+          end
+
+          current_resource.profile(current_profile)
+        end
+
+        def current_profile
+          all_profiles = get_installed_profiles
+
+          if all_profiles && all_profiles.key?("_computerlevel")
+            return all_profiles["_computerlevel"].find do |item|
+              item["ProfileIdentifier"] == new_profile_identifier
+            end
+          end
+          nil
+        end
+
+        def check_resource_semantics!
+          if action == :remove
+            if new_profile_identifier
+              unless !new_profile_identifier.nil? && !new_profile_identifier.end_with?(".mobileconfig") && /^\w+(?:(\.| )\w+)+$/.match(new_profile_identifier)
+                raise "when removing using the identifier property, it must match the profile identifier"
+              end
+            else
+              new_profile_name = new_resource.profile_name
+              unless !new_profile_name.end_with?(".mobileconfig") && /^\w+(?:(\.| )\w+)+$/.match(new_profile_name)
```

Comment:
This regex is used twice here, might be worth moving to a constant with some descriptive name.

Reply:
Actually these two conditionals are testing for the same thing on different values, so extracting that to a function with unit tests, etc might be a good idea.

Reply:
what would you call that conditional, because i can't actually tell what is going on there, the error text reads like we're comparing two values when we're just checking a single value?

Reply:
This:

`unless !new_profile_identifier.end_with?(".mobileconfig") && /^\w+(?:(\.| )\w+)+$/.match(new_profile_identifier)`

can be converted to this:

`if !( !new_profile_identifier.end_with?(".mobileconfig") && /^\w+(?:(\.| )\w+)+$/.match(new_profile_identifier) )`

Applying DeMorgan's law gives you:

`if new_profile_identifier.end_with?(".mobileconfig") || !/^\w+(?:(\.| )\w+)+$/.match(new_profile_identifier)`

So it raises an exception when:
* The string ends with `.mobileconfig`
* The string is not two or more groups of ascii characters separated by a space or `.`

It sounds like `.mobileconfig` files are specific to ios devices so excluding those makes sense. I don't understand why `"MyProfile"` or `"My      Profile"` are excluded.

In any case, I think the error message is trying to say something like `"When removing by resource name, the argument must be a valid profile identifier"` but I'm not sure why you'd want to proactively check for that rather than just allowing the underlying thing to fail.